File: alertmanager/telegram-bridge.js
const http = require('http');
const { spawn } = require('child_process');

const PORT = Number(process.env.TELEGRAM_BRIDGE_PORT || 9095);
const HOST = process.env.TELEGRAM_BRIDGE_HOST || '127.0.0.1';
const BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const CHAT_ID = process.env.TELEGRAM_CHAT_ID;

if (!BOT_TOKEN) {
  console.error('Missing TELEGRAM_BOT_TOKEN');
  process.exit(1);
}
if (!CHAT_ID) {
  console.error('Missing TELEGRAM_CHAT_ID');
  process.exit(1);
}

function escapeHtml(value) {
  return String(value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

function formatDate(value) {
  if (!value) return '';
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? escapeHtml(value) : escapeHtml(date.toISOString());
}

function formatKeyValues(values, skip = []) {
  return Object.entries(values || {})
    .filter(([k, v]) => !skip.includes(k) && v !== undefined && v !== null && v !== '')
    .map(([k, v]) => `${escapeHtml(k)}=${escapeHtml(v)}`)
    .join(' ');
}

function summarizeAlert(alert) {
  const labels = alert.labels || {};
  const annotations = alert.annotations || {};
  const status = escapeHtml(alert.status || 'firing');
  const name = escapeHtml(labels.alertname || 'Alert');
  const severity = escapeHtml(labels.severity || 'unknown');
  const summary = annotations.summary ? `\n<b>Summary:</b> ${escapeHtml(annotations.summary)}` : '';
  const description = annotations.description ? `\n<b>Description:</b> ${escapeHtml(annotations.description)}` : '';
  const impact = annotations.impact ? `\n<b>Impact:</b> ${escapeHtml(annotations.impact)}` : '';
  const action = annotations.action ? `\n<b>Action:</b> ${escapeHtml(annotations.action)}` : '';
  const runbook = annotations.runbook_url ? `\n<b>Runbook:</b> ${escapeHtml(annotations.runbook_url)}` : '';
  const dashboard = annotations.dashboard_url ? `\n<b>Dashboard:</b> ${escapeHtml(annotations.dashboard_url)}` : '';
  const valueBlock = annotations.value ? `\n<b>Value:</b> ${escapeHtml(annotations.value)}` : '';
  const startsAt = alert.startsAt ? `\n<b>Started:</b> ${formatDate(alert.startsAt)}` : '';
  const endsAt = alert.endsAt ? `\n<b>Ended:</b> ${formatDate(alert.endsAt)}` : '';
  const source = alert.generatorURL ? `\n<b>Source:</b> ${escapeHtml(alert.generatorURL)}` : '';
  const labelPairs = formatKeyValues(labels, ['alertname']);
  const labelBlock = labelPairs ? `\n<b>Labels:</b> ${labelPairs}` : '';
  return `• <b>${name}</b>\n<b>Status:</b> ${status}\n<b>Severity:</b> ${severity}${summary}${description}${impact}${action}${valueBlock}${runbook}${dashboard}${startsAt}${endsAt}${source}${labelBlock}`;
}

function buildMessage(payload) {
  const status = escapeHtml(payload.status || 'firing');
  const alerts = Array.isArray(payload.alerts) ? payload.alerts : [];
  const common = payload.commonLabels || {};
  const groupName = escapeHtml(common.alertname || (alerts[0] && alerts[0].labels && alerts[0].labels.alertname) || 'Alert');
  const severity = escapeHtml(common.severity || (alerts[0] && alerts[0].labels && alerts[0].labels.severity) || 'unknown');
  const groupLabels = formatKeyValues(payload.groupLabels || {});
  const groupBlock = groupLabels ? `\n<b>Group labels:</b> ${groupLabels}` : '';
  const prefix = status === 'resolved' ? '✅' : severity === 'critical' ? '🚨' : severity === 'warning' ? '⚠️' : '🔔';
  const body = alerts.slice(0, 10).map(summarizeAlert).join('\n\n');
  const truncated = alerts.length > 10 ? `\n\n…and ${alerts.length - 10} more alerts` : '';
  return `${prefix} <b>${groupName}</b>\n<b>Group status:</b> ${status}\n<b>Alert count:</b> ${alerts.length}${groupBlock}${body ? `\n\n${body}` : ''}${truncated}`;
}

function sendTelegramMessage(message) {
  const body = JSON.stringify({
    chat_id: CHAT_ID,
    text: message,
    parse_mode: 'HTML',
    disable_web_page_preview: true,
  });

  const child = spawn('curl', [
    '--silent',
    '--show-error',
    '--fail',
    '--max-time', '20',
    '-H', 'Content-Type: application/json',
    '-d', body,
    `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`,
  ]);

  let stdout = '';
  let stderr = '';
  child.stdout.on('data', (chunk) => {
    stdout += chunk;
  });
  child.stderr.on('data', (chunk) => {
    stderr += chunk;
  });
  child.on('close', (code) => {
    if (code === 0) {
      console.log('telegram_send_ok', stdout.trim() || '200');
    } else {
      console.error('telegram_send_failed', code, stderr.trim() || stdout.trim());
    }
  });
}

function handleWebhook(req, res, raw) {
  let payload;
  try {
    payload = JSON.parse(raw || '{}');
  } catch (err) {
    res.writeHead(400, { 'Content-Type': 'text/plain' });
    res.end('invalid json');
    return;
  }

  const message = buildMessage(payload);
  console.log('webhook_received', JSON.stringify({
    status: payload.status,
    alertCount: Array.isArray(payload.alerts) ? payload.alerts.length : 0,
    groupLabels: payload.groupLabels || {},
    commonLabels: payload.commonLabels || {},
  }));
  sendTelegramMessage(message);

  res.writeHead(200, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ ok: true }));
}

const server = http.createServer((req, res) => {
  if (req.method === 'GET' && req.url === '/health') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ ok: true }));
    return;
  }

  if (req.method === 'POST' && req.url === '/alert') {
    let raw = '';
    req.on('data', (chunk) => {
      raw += chunk;
      if (raw.length > 1024 * 1024) {
        req.destroy(new Error('payload too large'));
      }
    });
    req.on('end', () => handleWebhook(req, res, raw));
    return;
  }

  res.writeHead(404, { 'Content-Type': 'text/plain' });
  res.end('not found');
});

server.listen(PORT, HOST, () => {
  console.log(`telegram_bridge_listening ${HOST}:${PORT}`);
});

File: alertmanager/telegram-bridge.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json
import os
import subprocess
import sys
from datetime import datetime, timezone

HOST = os.environ.get('TELEGRAM_BRIDGE_HOST', '127.0.0.1')
PORT = int(os.environ.get('TELEGRAM_BRIDGE_PORT', '9095'))
BOT_TOKEN = os.environ.get('TELEGRAM_BOT_TOKEN')
CHAT_ID = os.environ.get('TELEGRAM_CHAT_ID')

if not BOT_TOKEN:
    print('Missing TELEGRAM_BOT_TOKEN', file=sys.stderr)
    sys.exit(1)
if not CHAT_ID:
    print('Missing TELEGRAM_CHAT_ID', file=sys.stderr)
    sys.exit(1)


def esc(value):
    return (str(value)
        .replace('&', '&amp;')
        .replace('<', '&lt;')
        .replace('>', '&gt;')
        .replace('"', '&quot;'))

def fmt_time(value):
    if not value:
        return 'n/a'
    try:
        normalized = str(value).replace('Z', '+00:00')
        dt = datetime.fromisoformat(normalized)
        return dt.astimezone(timezone.utc).strftime('%Y-%m-%d %H:%M:%S UTC')
    except Exception:
        return esc(value)

def pick_labels(labels):
    keys = ['environment', 'cluster', 'service', 'job', 'instance', 'container', 'severity']
    return ' '.join(
        f"{esc(k)}={esc(labels[k])}"
        for k in keys
        if labels.get(k)
    )


def summarize_alert(alert):
    labels = alert.get('labels', {})
    annotations = alert.get('annotations', {})
    status = esc(alert.get('status', 'firing'))
    name = esc(labels.get('alertname', 'Alert'))
    severity = esc(labels.get('severity', 'unknown'))
    starts_at = esc(fmt_time(alert.get('startsAt')))
    ends_at = esc(fmt_time(alert.get('endsAt')))
    summary = f"\n<b>Summary:</b> {esc(annotations['summary'])}" if annotations.get('summary') else ''
    description = f"\n<b>Description:</b> {esc(annotations['description'])}" if annotations.get('description') else ''
    impact = f"\n<b>Impact:</b> {esc(annotations['impact'])}" if annotations.get('impact') else ''
    action = f"\n<b>Action:</b> {esc(annotations['action'])}" if annotations.get('action') else ''
    value = f"\n<b>Value:</b> {esc(alert['value'])}" if alert.get('value') else ''
    generator = f"\n<b>Source:</b> {esc(alert['generatorURL'])}" if alert.get('generatorURL') else ''
    labels_text = pick_labels(labels)
    label_block = f"\n<b>Labels:</b> {labels_text}" if labels_text else ''
    resolved_at = f"\n<b>Resolved at:</b> {ends_at}" if status == 'resolved' else ''
    return (
        f"• <b>{name}</b>\n<b>Status:</b> {status}\n<b>Severity:</b> {severity}"
        f"\n<b>Started at:</b> {starts_at}{resolved_at}{summary}{description}"
        f"{impact}{action}{value}{label_block}{generator}"
    )


def build_message(payload):
    status = esc(payload.get('status', 'firing'))
    alerts = payload.get('alerts', []) or []
    common = payload.get('commonLabels', {}) or {}
    first = alerts[0] if alerts else {}
    first_labels = first.get('labels', {})
    group_name = esc(common.get('alertname') or first_labels.get('alertname') or 'Alert')
    severity = esc(common.get('severity') or first_labels.get('severity') or 'unknown')
    prefix = '✅' if status == 'resolved' else '🚨' if severity == 'critical' else '⚠️' if severity == 'warning' else '🔔'
    body = '\n\n'.join(summarize_alert(a) for a in alerts[:10])
    truncated = f"\n\n…and {len(alerts) - 10} more alerts" if len(alerts) > 10 else ''
    env = esc(common.get('environment', 'unknown'))
    cluster = esc(common.get('cluster', 'unknown'))
    return (
        f"{prefix} <b>{group_name}</b>\n<b>Group status:</b> {status}"
        f"\n<b>Severity:</b> {severity}\n<b>Environment:</b> {env}"
        f"\n<b>Cluster:</b> {cluster}\n<b>Alert count:</b> {len(alerts)}"
    ) + (f"\n\n{body}" if body else '') + truncated


def send_telegram(message):
    payload = json.dumps({
        'chat_id': CHAT_ID,
        'text': message,
        'parse_mode': 'HTML',
        'disable_web_page_preview': True,
    })
    result = subprocess.run([
        'curl', '--silent', '--show-error', '--fail', '--max-time', '20',
        '-H', 'Content-Type: application/json',
        '-d', payload,
        f'https://api.telegram.org/bot{BOT_TOKEN}/sendMessage'
    ], capture_output=True, text=True)
    if result.returncode == 0:
        print('telegram_send_ok', result.stdout.strip() or '200', flush=True)
        return True
    print('telegram_send_failed', result.returncode, result.stderr.strip() or result.stdout.strip(), flush=True)
    return False


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == '/health':
            self.send_response(200)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(b'{"ok":true}')
            return
        if self.path == '/metrics':
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain; version=0.0.4')
            self.end_headers()
            self.wfile.write(b'telegram_bridge_up 1\n')
            return
        self.send_response(404)
        self.end_headers()

    def do_POST(self):
        if self.path != '/alert':
            self.send_response(404)
            self.end_headers()
            return
        length = int(self.headers.get('Content-Length', '0'))
        raw = self.rfile.read(length)
        try:
            payload = json.loads(raw.decode('utf-8') if raw else '{}')
        except Exception:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b'invalid json')
            return
        print('webhook_received', json.dumps({
            'status': payload.get('status'),
            'alertCount': len(payload.get('alerts', []) or []),
            'groupLabels': payload.get('groupLabels', {}),
            'commonLabels': payload.get('commonLabels', {}),
        }), flush=True)
        send_telegram(build_message(payload))
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(b'{"ok":true}')

    def log_message(self, format, *args):
        return


if __name__ == '__main__':
    print(f'telegram_bridge_listening {HOST}:{PORT}', flush=True)
    HTTPServer((HOST, PORT), Handler).serve_forever()
